rasa/telemetry.py

```diff
@@ -0,0 +1,506 @@
+        print(json.dumps(payload, indent=2))
+        return
+
+    write_key = telemetry_write_key()
+    if not write_key:
+        # If TELEMETRY_WRITE_KEY is empty or `None`, telemetry has not been
+        # enabled for this build (e.g. because it is running from source)
+        logger.debug("Skipping request to external service: telemetry key not set.")
+        return
+
+    headers = segment_request_header(write_key)
+
+    async with aiohttp.ClientSession() as session:
+        async with session.post(
+            SEGMENT_ENDPOINT,
+            headers=headers,
+            json=payload,
+            timeout=TELEMETRY_HTTP_TIMEOUT,
+        ) as resp:
+            # handle different failure cases
+            if resp.status != 200:
+                response_text = await resp.text()
+                logger.debug(
+                    f"Segment telemetry request returned a {resp.status} response."
+                    f"Body: {response_text}"
+                )
+            else:
+                data = await resp.json()
+                if not data.get("success"):
+                    logger.debug(
+                        f"Segment telemetry request returned a failure."
+                        f"Response: {data}"
+                    )
+
+
+def _project_hash() -> Text:
+    """Create a hash for the project in the current working directory.
+
+    Returns:
+        project hash
+    """
+    try:
+        remote = check_output(  # skipcq:BAN-B607,BAN-B603
+            ["git", "remote", "get-url", "origin"], stderr=STDOUT
+        )
+        return hashlib.sha256(remote).hexdigest()
+    except (CalledProcessError, OSError):
+        working_dir = Path(os.getcwd()).absolute()
+        return hashlib.sha256(str(working_dir).encode("utf-8")).hexdigest()
+
+
+def with_default_context_fields(
+    context: Optional[Dict[Text, Any]] = None,
+) -> Dict[Text, Any]:
+    """Return a new context dictionary that contains the default field values merged
+    with the provided ones. The default fields contain only the OS information for now.
+
+    Args:
+        context: Context information about the event.
+
+    Return:
+        A new context.
+    """
+    context = context or {}
+
+    return {**_default_context_fields(), **context}
+
+
+@functools.lru_cache()
+def _default_context_fields() -> Dict[Text, Any]:
+    """Return a dictionary that contains the default context values.
+
+    Return:
+        A new context containing information about the runtime environment.
+    """
+    import tensorflow as tf
+
+    return {
+        "os": {"name": platform.system(), "version": platform.release(),},
+        "ci": in_continuous_integration(),
+        "project": _project_hash(),
+        "python": sys.version.split(" ")[0],
+        "rasa_open_source": rasa.__version__,
+        "gpu": len(tf.config.list_physical_devices("GPU")),
+        "cpu": multiprocessing.cpu_count(),
+    }
+
+
+@ensure_telemetry_enabled
+async def track(
```

Comment:
The way `track` is implemented now, callers need to use `await track(...)`. This implies that, although other coroutines can run while the HTTP request is being sent, the caller of `track` will still need to wait for the request to end until they can continue with their code. For example, this happens on `await track_model_training()`.

My idea was to implement `track()` without the `async` keyword, but still have it use the event loop to call `_send_event`, but in an optional fire-and-forget fashion:
```python
def track(..., async_context: bool = True):
    coro = _send_event(...)
    
    if async_context:
        asyncio.ensure_future(coro)  # Coroutine will be added to the loop, no need to await its result
    else:
        asyncio.get_event_loop().run_until_complete(coro)  # Block until it's finished
```

So we would use `async_context=True` in contexts where there is already an event loop running, and we want to fire-and-forget the request. We would use `False` in contexts where there is no event loop running, like in `track_telemetry_disabled`.

The disadvantage of using fire-and-forget is that in some cases, the event loop could be closed before the coroutine is run. I can see this happening for example when we finish training the model and want to send an event right at the end.

All of this only applies if we want to be very conscious of potential delays by telemetry requests though. If we don't care about those then all of this isn't necessary and we can keep a simple `await`able interface.

Reply:
yes that is a good point, I changed it to run in the background in a slightly different way, using `enusre_future` in the event specific telemetry functions